### PR TITLE
ENH add openssl-like version update code

### LINKED ISSUE #1223: "lettered" version number don't work out of the bot with the updating bot
Openssl and tzdata are two examples of non-pep440 version numbers that will be ignored at the moment and won't get updates.

xref.: https://github.com/conda-forge/tzdata-feedstock/issues/4

### PR DESCRIPTION
**Pull request by beckermr**
This PR adds code to get openssl-like versions.

closes #1223

### CODE REVIEW

**Comment by beckermr** [comment]
@ocefpaf @CJ-Wright @mariusvniekerk @isuruf what other packages use openssl like versions?

**Comment by CJ-Wright** [comment]
Can you narrow that down a bit more? Is this just letters at the end?

**Comment by CJ-Wright** [comment]
{'cddlib': '1!0.94j',
 'clrng': '1.0.0b',
 'fig2dev': '3.2.6a',
 'jpeg': '9d',
 'libmad': '0.15.1b',
 'lrslib': '70.a',
 'lsst_dashboard': '0.0.2a',
 'mad': '0.15.1b',
 'multicore-tsne': '0.1_d4ff4aab',
 'odm2rest': '0.1.0a',
 'openpmd-api': '0.12.0a',
 'openssl': '1.1.1i',
 'pyorgmode': '0.03c',
 'tzcode': '2020a',
 'tzdata': '2020d',
 'wgrib': '1.8.1.2c',
 'xfig': '3.2.6a'}

**Comment by beckermr** [comment]
Some of those might be dev versions though?

**Comment by CJ-Wright** [comment]
Maybe? I'm not really in a position to know. I'm also not certain that the dev/non-dev is a great dividing line either, black isn't on that list, but should it be?

**Comment by beckermr** [comment]
black is shipping a legit dev version. my understanding is that one day they will not do that

**Comment by CJ-Wright** [comment]
Maybe a different framing is "if you are already shipping a dev version we'll help you ship dev versions until you aren't shipping dev versions". The set above is what the bot has for the latest versions of things.

**Comment by beckermr** [comment]
That's an interesting idea. I am going to punt on it for now and instead only include openssl jpeg tzcode and tzdata.

**Comment by CJ-Wright** [comment]
Things that most likely should also be on that list:
- [cddlib](https://github.com/cddlib/cddlib/releases)
- [jpeg](http://www.ijg.org/files/)


It seems that pyorgmode is defunct

**Comment by beckermr** [comment]
OK @CJ-Wright I fixed several bugs in the feedstock parser and added a better way to do openssl-like versions. I think this one is ready to go!

File: conda_forge_tick/feedstock_parser.py [hunk 1]
```diff
@@ -164,6 +164,12 @@ def populate_feedstock_attributes(
     if isinstance(meta_yaml, Response):
         sub_graph["bad"] = f"make_graph: {meta_yaml.status_code}"
         return sub_graph
+
+    # strio out old keys - this removes old platforms when one gets disabled
```
Comment (beckermr) [inline conda_forge_tick/feedstock_parser.py:168]:
```suggestion
    # strip out old keys - this removes old platforms when one gets disabled
```

**Comment by beckermr** [comment]
cc @ocefpaf This should fix version updates for tzdata.

File: conda_forge_tick/feedstock_parser.py [hunk 2]
```diff
@@ -285,8 +291,17 @@ def populate_feedstock_attributes(
...
     sub_graph["req"] = req
 
+    # set name and version
     keys = [("package", "name"), ("package", "version")]
     missing_keys = [k[1] for k in keys if k[1] not in yaml_dict.get(k[0], {})]
+    for k in keys:
+        if k[1] not in missing_keys:
+            sub_graph[k[1]] = yaml_dict[k[0]][k[1]]
+
+    # set the url and hash
```
Comment (CJ-Wright) [inline conda_forge_tick/feedstock_parser.py:301]: Where are we getting these pieces of information now, or are they no longer parts of the metadata?

Comment (beckermr) [reply conda_forge_tick/feedstock_parser.py:301]: They come from the metadata in the rendered recipe.

File: tests/test_upstream_versions.py [hunk 3]
```diff
@@ -184,3 +184,25 @@ def test_latest_version(inp, ver, source, urls, requests_mock, tmpdir):
...
+        ("8.1.5", ["8.1.6", "8.2.0", "9.0.0"]),
+        ("8_1", ["8_2", "9_0"]),
+        ("8_1_5", ["8_1_6", "8_2_0", "9_0_0"]),
+        ("8-1", ["8-2", "9-0"]),
+        ("8-1-5", ["8-1-6", "8-2-0", "9-0-0"]),
+        ("8.1-10", ["8.1-11", "8.2-0", "9.0-0"]),
+        ("8.1_10", ["8.1_11", "8.2_0", "9.0_0"]),
+        ("10.8.1-10", ["10.8.1-11", "10.8.2-0", "10.9.0-0", "11.0.0-0"]),
+        ("10-8.1_10", ["10-8.1_11", "10-8.2_0", "10-9.0_0", "11-0.0_0"]),
+        ("1.1.1a", ["1.1.1b", "1.1.2a", "1.2.0a", "2.0.0a"]),
```
Comment (CJ-Wright) [inline tests/test_upstream_versions.py:202]: Should we also handle cases that remove the suffix letter? For instance upgrading from alpha to a subsequent full release

File: conda_forge_tick/feedstock_parser.py (hunk 2 above)
Comment (CJ-Wright) [reply conda_forge_tick/feedstock_parser.py:301]: Does the `hash_type` key still exist at top level?

Comment (beckermr) [reply conda_forge_tick/feedstock_parser.py:301]: yes

**Comment by beckermr** [comment]
we good to merge?

**Comment by CJ-Wright** [comment]
sure, can you babysit?

**Comment by beckermr** [comment]
yep!

**Pull request merged**